contrib/opentimelineio_contrib/adapters/xges.py

```diff
@@ -190,507 +300,925 @@ def to_rational_time(self, ns_timestamp):
-            if type(tmp_dur) is int:
-                duration = tmp_dur
-            else:
-                print("WARNING: read a duration that is not an int")
+        duration = self._get_from_properties(
+            asset, "duration", "guint64")
 
-        available_range = otio.opentime.TimeRange(
-            start_time=self.to_rational_time(0),
-            duration=self.to_rational_time(duration)
-        )
+        if duration is None:
+            available_range = None
+        else:
+            available_range = otio.opentime.TimeRange(
+                start_time=self.to_rational_time(0),
+                duration=self.to_rational_time(duration)
+            )
         ref = otio.schema.ExternalReference(
-            target_url=asset.get("id"),
+            target_url=asset_id,
             available_range=available_range
         )
-
-        ref.metadata[META_NAMESPACE] = {
-            "properties": asset.get("properties"),
-            "metadatas": asset.get("metadatas"),
-        }
-
+        self._add_properties_and_metadatas_to_otio(ref, asset)
         return ref
 
     # --------------------
     # search helpers
     # --------------------
     def _asset_by_id(self, asset_id, asset_type):
-        return self.xges_xml.find(
-            "./project/ressources/asset[@id='{}'][@extractable-type-name='{}']".format(
-                asset_id, asset_type)
+        return self._findonly(
+            self.ges_xml,
+            "./project/ressources/asset[@id='{}']"
+            "[@extractable-type-name='{}']".format(
+                asset_id, asset_type),
+            allow_none=True
         )
 
     def _timeline_element_by_name(self, timeline, name):
-        for clip in timeline.findall("./layer/clip"):
-            if self._get_from_properties(clip, "name") == name:
+        for clip in self._findall(timeline, "./layer/clip"):
+            if self._get_from_properties(clip, "name", "string") == name:
                 return clip
 
         return None
 
 
 class XGESOtio:
+    """
+    This object is responsible for converting an otio timeline into an
+    xGES project.
+    """
 
-    def __init__(self, input_otio):
-        self.container = input_otio
-        self.rate = 25.0
+    def __init__(self, input_otio=None):
+        if input_otio is not None:
+            self.timeline = input_otio.deepcopy()
+        else:
+            self.timeline = None
+        self.all_names = set()
+
+    @staticmethod
+    def to_gstclocktime(otio_time):
+        """
+        Convert an otio time into a GstClockTime. If a RationalTime is
+        received, returns a single int representing the time in
+        nanoseconds. If a TimeRange is received, returns a tuple of the
+        start_time and duration as ints representing the times in
+        nanoseconds.
+        """
+        if isinstance(otio_time, otio.opentime.RationalTime):
+            return int(otio_time.value_rescaled_to(1) * GST_SECOND)
```

Comment:
Also there is `otio.opentime.to_seconds(RationalTime) -> Double`

Reply:
thanks!